### DISCUSSION TITLE
How to build a universal MachO library 'amd64+arm64' for MacOS?

### DISCUSSION THREAD

**Discussion by jugglingjsons** [quaadgras/graphics.gd#31] [Q&A] [answered]
I am having issues with my export on macOS - not sure what I am doing wrong, I was wondering if we could have a working example as title says.
I would need the help of someone more knowledgable to get that working, since I am not getting anywhere myself though.

**Comment by Splizard** [comment]
Are you trying to build for other platforms and architectures? ie. Windows and Linux from Mac. 

And/or are you having issues with exporting the project for Mac?

**Reply by jugglingjsons** [reply]
I did indeed run into issues.
I think the correct format for the extension setup is: (this is what my exports work with)

> macos.debug   = "darwin_${arch}.dylib"
> macos.release    = "darwin_${arch}.dylib"

whereas the current is set to: (with this the export tells me there is no such file / file not found for platform)

> macos.x86_64   = "darwin_amd64.dylib"
>macos.arm64    = "darwin_arm64.dylib"

I have an additional question however. The repo readme states you may have some ideas as to how to implement wasm export support, on the other hand the documentation says there is no plan to implement that.
The question is two-fold:
1. Is wasm export planned ?
2. Can you explain a bit how you see it? I could help with that.

**Reply by Splizard** [reply]
On the WASM question, we need a javascript-accessible interface to the GDExtension API. Once we have this, it should be fairly straightforward to support it. At the moment, GDExtension is only available on web via the Emscripten ABI (so pretty much restricted to C/C++).

**Reply by Splizard** [reply]
On the export issue, is this because for MacOS, Godot is trying to create a universal .MachO ? With amd64+arm64 packed into the same binary? If this is the case you will need to use GOARCH variables to build both libraries.

**Reply by jugglingjsons** [reply]
Exactly that - universal builds, I have tried exporting both binaries separately for now. I am not sure if there is any templating possible for the extension file.
I mean I have run separate exports with the go file compiled for specific arch.
Both builds worked fine then - yet to check if I can somehow make it work with both binaries built in the same export.

**Reply by Splizard** [reply]
I believe there is mac os tooling to merge two binaries together for this. Not sure how to signal this to Godot, the values used in the extension file are a bit bespoke. Essentially Godot looks for a whole bunch of matching values, to get the best match, these values aren't explicitly documented or specified.

**Reply by jugglingjsons** [reply, marked as answer]
You are correct. I've verified it over the weekend and using:(same versions run on intel mac and m1, build on m1)

`env CGO_ENABLED=1 CGO_CFLAGS=-O3 GOARCH=amd64 go build -o darwin_amd64.dylib -buildmode=c-shared`

and 

`env CGO_ENABLED=1 CGO_CFLAGS=-O3 GOARCH=arm64 go build -o darwin_amd64.dylib -buildmode=c-shared`

and bundling it later on with :

`lipo -create <pathToIntel.a> <pathToArm.a> -output <pathToUniversal.a>`

Yields a working universal build.
If you're open to PRs, I will add that to the documentation with links as to why and such.

**Reply by Splizard** [reply]
Thanks, I'm certainly keen to support this, if you were looking to contribute here, perhaps the `gd` command could handle this automatically?

**Reply by jugglingjsons** [reply]
Took me a while, but opened an initial [PR](https://github.com/grow-graphics/gd/pull/47), let's continue the discussion there ;)
